smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/lang/DocumentationPreprocessor.kt

```diff
@@ -21,17 +26,226 @@ class DocumentationPreprocessor : KotlinIntegration {
+                    // that no information is destroyed. The worst outcome is that some seemingly nonsense HTML tag is
+                    // injected into the output, which a reader can reasonably ignore.
+                    text += "<${node.nodeName()}>"
+                }
+            }
+        }
+
+        override fun tail(node: Node, depth: Int) {
+            when (node.nodeName()) {
+                "p", "div", "dd" -> {
+                    val nextSibling = node.nextSibling()
+                    text += when {
+                        // break to give the upcoming list a new line
+                        (nextSibling != null && nextSibling.isList()) -> "\n"
+                        // if we're inside a list, the outer list item will close out the line for us
+                        (node.hasAncestor(Node::isList)) -> ""
+                        // all other cases: this is a standalone "text block" which should be displayed as its own
+                        // paragraph
+                        else -> "\n\n"
+                    }
+                }
+                "a" -> {
+                    writeBufferedAnchor()
+                }
+                "ul", "ol" -> {
+                    sublistDedent()
+                    text += if (node.parent()?.nodeName() == "body") "\n\n" else ""
+                }
+                "code", "pre" -> {
+                    text += PREFORMAT_MARKER
+                }
+                "b", "strong" -> {
+                    text += BOLD_MARKER
+                }
+                "i", "em" -> {
+                    text += ITALIC_MARKER
+                }
+                "li", "fullname", "dt" -> {
+                    text += "\n"
+                }
+            }
+        }
+
+        private fun writeBufferedAnchor() {
+            // Model docs will sometimes contain an anchor without the href. At that point there's no real way of
+            // knowing to what it refers, nor can we guarantee a valid link just by bracketing it.
+            text += if (bufferedAnchorHref != "") {
+                "[$bufferedAnchorText]($bufferedAnchorHref)"
+            } else {
+                bufferedAnchorText
+            }
+
+            bufferedAnchorHref = ""
+            bufferedAnchorText = ""
+        }
+
+        private fun sublistIndent() {
+            listPrefix += SUBLIST_INDENT
+        }
+
+        private fun sublistDedent() {
+            listPrefix = listPrefix.dropLast(SUBLIST_INDENT.length)
+        }
+    }
+}
+
+/**
+ * Jsoup will preserve newlines between elements as blank text nodes. These have zero bearing on the content of the
+ * document to begin with and only serve to complicate traversal.
+ */
+private fun Node.stripBlankTextNodes() {
+    if (this is TextNode && this.isBlank) {
+        this.remove()
+        return
+    }
+
+    val childNodes = this.childNodes()
+    if (childNodes.isNotEmpty()) {
+        childNodes.forEach(Node::stripBlankTextNodes)
+    }
```

Comment:
**Nit**: Checking if a collection is empty before iterating it is unnecessary. This can be simplified to:

```kotlin
childNodes().forEach(Node::stripBlankTextNodes)
```